facilitator/src/transport.rs

```diff
@@ -88,11 +100,314 @@ impl Transport for NullTransport {
+                        )
+                        .into(),
+                    ),
+                    ..Default::default()
+                }),
+            )
+            .map_err(|e| {
+                // Clean up botched uploads
+                if let Err(abort) = self.abort_upload() {
+                    return std::io::Error::new(
+                        std::io::ErrorKind::Other,
+                        format!(
+                            "error uploading part: {:?}\n\tAND failure aborting: {:?}",
+                            e, abort
+                        ),
+                    );
+                }
+                std::io::Error::new(
+                    std::io::ErrorKind::Other,
+                    format!("error uploading part: {:?}", e),
+                )
+            })?;
+
+        let completed_part = CompletedPart {
+            e_tag: Some(upload_output.e_tag.ok_or_else(|| {
+                // This error seems improbable but the e_tag field is an
+                // Optional so we have to do this dance.
+                if let Err(abort) = self.abort_upload() {
+                    return std::io::Error::new(
+                        std::io::ErrorKind::Other,
+                        format!(
+                            "no ETag in UploadPartOutput AND failure aborting: {:?}",
+                            abort
+                        ),
+                    );
+                }
+                std::io::Error::new(
+                    std::io::ErrorKind::Other,
+                    "no ETag in UploadPartOutput".to_owned(),
+                )
+            })?),
+            part_number: Some(part_number),
+        };
+        self.completed_parts.push(completed_part);
+        Ok(())
+    }
+
+    fn abort_upload(&mut self) -> std::io::Result<()> {
+        // There's nothing useful in the output so discard it
+        self.runtime
+            .block_on(
+                self.client
+                    .abort_multipart_upload(AbortMultipartUploadRequest {
+                        bucket: self.bucket.to_string(),
+                        key: self.key.to_string(),
+                        upload_id: self.upload_id.clone(),
+                        ..Default::default()
+                    }),
+            )
+            .map_err(|e| {
+                std::io::Error::new(
+                    std::io::ErrorKind::Other,
+                    format!("error completing upload: {:?}", e),
+                )
+            })?;
+        Ok(())
+    }
+}
+
+impl Write for MultipartUploadWriter {
+    fn write(&mut self, buf: &[u8]) -> std::io::Result<usize> {
+        // Write into memory buffer, and upload to S3 if we have accumulated
+        // enough content.
+        self.buffer.extend_from_slice(buf);
+        if self.buffer.len() >= self.buffer_capacity {
+            self.upload_part()?;
+        }
+
+        Ok(buf.len())
+    }
+
+    fn flush(&mut self) -> std::io::Result<()> {
```

Comment:
One consequence of this implementation is that callers _must_ call `std::io::Write::flush` on the `Write` they get or the upload will (1) never be completed and (2) we will be billed indefinitely for any upload parts Amazon is holding on to. We control all the callers in this same codebase, so we can make sure we flush() appropriately (hence the changes I made in `batch.rs`) but I wonder if `MultipartUploadWriter` could defend itself against this better. For instance, if we added an `std::ops::Drop` implementation that panics if `self.completed_parts` is non-empty, to help programmers detect what I think is a programming error rather than something that could reasonably happen at runtime.

Reply:
Rust's type system has no way of "preventing a type from being dropped" and thus forcing you to consume it somehow, so that's not really possible. The closest approximation to that is to have a function accepting a closure, which acquires the resource, runs the closure and then immediately cleans up the resource afterwards. This might be something we can do here but it would require a lot of changes and would probably be awkward to use.

Now of course you're free to impl `Drop` and then make a blocking call to the network to finalize the chunk, but that will not extend to async network calls, so it feels like a bad solution to have that be the main API and not just a fallback.

In addition, using `flush` on the writer (as you're doing now) is also a bad option, because it means that you're breaking the contract from the trait that `flush` can be called multiple times.

So in the end I think a reasonable thing to do is have a dedicated method in `MultipartUploadWriter` to finish the upload, and another to cancel it. As for the cleanup, panicing in `Drop` if none of the aforementioned methods were called is a possibility, but I don't know if it's actually a good idea in production. I suppose there are no cases in which we'd rather execution finish with a forgotten file which was not written (rather than aborting entirely), so this might be valid for our use-case (but probably not others, which is fine). An alternative is logging an error and then, in the background, calling the cancellation method as a last-ditch effort.

As for encouraging users to not forget to call the functions in the first place, a good way of doing this is to make the finalization call return some value you're going to need later in the program. I don't know if there's anything that would fit the bill here, but it's something to do if you can think of something that'd fit.

Regarding the costs of not finalizing an upload, we should set a [bucket lifecycle policy](https://docs.aws.amazon.com/AmazonS3/latest/dev/mpuoverview.html#mpu-stop-incomplete-mpu-lifecycle-config) which will expire unfinished uploads after a certain amount of time. We'd need to do this either way because there's always the chance that your machine/process randomly die at any point during the upload.

Reply:
Ooh, good point about the `flush` contract. The only problem with extending the `MultipartUploadWriter` method surface as you describe is that `Transport::put` would no longer return an `std::io::Write` but we control all the callers so that's not so bad.